scripts/test-new-tests.mjs

```diff
@@ -0,0 +1,132 @@
+// @ts-check
+import fs from 'fs/promises'
+import execa from 'execa'
+import path from 'path'
+
+async function main() {
+  let eventData = {}
+
+  /** @type import('execa').Options */
+  const EXECA_OPTS = { shell: true }
+  /** @type import('execa').Options */
+  const EXECA_OPTS_STDIO = { ...EXECA_OPTS, stdio: 'inherit' }
+
+  try {
+    eventData =
+      JSON.parse(
+        await fs.readFile(process.env.GITHUB_EVENT_PATH || '', 'utf8')
+      )['pull_request'] || {}
+  } catch (_) {}
+
+  // detect changed test files
+  const branchName =
+    eventData?.head?.ref ||
+    process.env.GITHUB_REF_NAME ||
+    (await execa('git rev-parse --abbrev-ref HEAD', EXECA_OPTS)).stdout
+
+  const remoteUrl =
+    eventData?.head?.repo?.full_name ||
+    process.env.GITHUB_REPOSITORY ||
+    (await execa('git remote get-url origin', EXECA_OPTS)).stdout
+
+  const isCanary =
+    branchName.trim() === 'canary' && remoteUrl.includes('vercel/next.js')
+
+  if (isCanary) {
+    console.error(`Skipping flake detection for canary`)
+    return
+  }
+
+  try {
+    await execa('git remote set-branches --add origin canary', EXECA_OPTS_STDIO)
+    await execa('git fetch origin canary --depth=20', EXECA_OPTS_STDIO)
+  } catch (err) {
+    console.error(await execa('git remote -v', EXECA_OPTS_STDIO))
+    console.error(`Failed to fetch origin/canary`, err)
+  }
+
+  const changesResult = await execa(
+    `git diff origin/canary --name-only`,
+    EXECA_OPTS
+  ).catch((err) => {
+    console.error(err)
+    return { stdout: '', stderr: '' }
+  })
+  console.error(
+    {
+      branchName,
+      remoteUrl,
+      isCanary,
+    },
+    `\ngit diff:\n${changesResult.stderr}\n${changesResult.stdout}`
+  )
+  const changedFiles = changesResult.stdout.split('\n')
+  console.log('detected tests:', changedFiles)
```

Comment:
since this is all changedFiles, should we update the log? Or maybe only log once we've matched `.test.*` files?

Reply:
Good point, moved it down a bit to after full detection